New simplified uploader

tvdeyen opened the pull request:
Removed the extra dialog for uploading images. Instead the upload button is now a globally usable file select field and the whole associated file list (image archive, attachments) is now a drop target.

Moves the upload progress to the bottom of the screen.

Happy drag'n'drop uploading everywhere in your Rails App |o/

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://simplified-uploader.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin' target='_blank'>Open stage</a> or <a href='http://simplified-uploader.alchemycms-alchemy-cms-42e8f4fc945384e0f6b5.ttrcloud.com/admin/teatro.log' target='_blank'>view logs</a>.

tvdeyen commented on app/views/alchemy/admin/attachments/_archive_overlay.html.erb:
```diff
@@ -1,21 +1,18 @@
-        title: Alchemy.t(:upload_file),
-        class: 'icon_button'
+  <% if can? :create, Alchemy::Attachment %>
+    <%= render 'alchemy/admin/partials/upload_button',
```
Maybe we should move the uploader partials into its own folder? `alchemy/admin/uploader/button`

mamhoff replied:
Yes!

tvdeyen commented on app/views/alchemy/admin/attachments/_archive_overlay.html.erb:
```diff
@@ -1,21 +1,18 @@
+      dropzone: '#assign_file_list',
+      form_object: Alchemy::Attachment.new,
+      file_attribute: 'file',
+      while_assigning: true,
```
This shoul be better `in_dialog`

mamhoff replied:
:+1:

tvdeyen commented on app/views/alchemy/admin/partials/_upload_button.html.erb:
```diff
@@ -0,0 +1,31 @@
+<%= form_for [alchemy, :admin, form_object], html: {multipart: true, class: 'upload-button'} do |f| %>
```
If we pass the `alchemy` url proxy here, the button is not usable outside of Alchemy

mamhoff replied:
Let's keep it as-is, if people need this to be used outside of Alchemy, let's add that complexity later.

tvdeyen replied:
It works without the proxy as well and removing it will actually remove complexity.

mamhoff replied:
:+1:

mamhoff commented on app/assets/javascripts/alchemy/alchemy.uploader.js.coffee:
```diff
@@ -19,28 +19,21 @@ Alchemy.Uploader = (settings) ->
-      $.merge(form_data, settings.post_params)
-      form_data
+    maxNumberOfFiles: Alchemy.uploader_defaults.file_upload_limit
+    maxFileSize: Alchemy.uploader_defaults.file_size_limit * 1000000
```
Is this a magic number? What does it mean? :)

tvdeyen replied:
1 Megabyte in JavaScript

mamhoff replied:
OK, cool!

mamhoff commented on app/views/alchemy/admin/pictures/_filter_and_size_bar.html.erb:
```diff
@@ -1,27 +1,20 @@
+        size: params[:size],
+        filter: 'last_upload',
+        content_id: @content.try(:id),
+        element_id: @element.try(:id)
```
Is there any way to reduce the number of local assigns we have to pass into this partial?

tvdeyen replied:
Thought a lot about removing them, but as you see in the partial, they are all necessary. And the `redirect_url` here is current API of the pictures controller. We could reduce the amount of parameters needed, but that is out of the scope of this PR

mamhoff replied:
:+1:

tvdeyen replied:
The new interface is much cleaner. Had to change the translation key and the configuration, though.

mamhoff commented on spec/controllers/alchemy/admin/attachments_controller_spec.rb:
```diff
@@ -105,23 +86,6 @@ module Alchemy
-            expect(Content).to receive(:find_by).and_return(content)
-          end
-
-          it "assigns lots of instance variables" do
```
:shipit:

mamhoff commented:
The specs are failing on Ruby 2.2+ and Postgres. I suspect intermittent failures; have re-triggered the builds for now. Do you want to look into those as well?

tvdeyen commented:
I already restarted them yesterday and it still breaks. Will check if this happens on master as well. `acts_as_list` was updated. Maybe that's the issue.

mamhoff commented:
Generally, this is a wonderful change (yay, -500 lines!)

tvdeyen commented:
@mamhoff I simplified the simplified upload button

mamhoff commented:
I saw the change with `in_dialog` vs. `while_assigning`, and also the ones where the button partial now has default values. Anything I missed? This is great anyways :)

mamhoff commented on config/locales/alchemy.es.yml:
```diff
@@ -566,8 +553,10 @@ es:
+    uploader:
+      label:
+        alchemy/attachment: Subir archivo(s)
+        alchemy/picture: Subir imagen(es)
```
👍

mamhoff commented on lib/alchemy/upgrader/three_point_three.rb:
```diff
@@ -37,6 +37,12 @@ def alchemy_3_3_todos
+
+The name of the model is now namespaced. `alchemy/pictures` instead of just `pictures`.
+Please ensure, to copy the new setting from the `config.yml.defaults` file.
+
```
Ah ok. I think this is a manageable change, even for 3.3.

tvdeyen commented:
@mamhoff Codeclimate complained duplicated code. So, I extracted the uploader JSON responses into a module. Therefore I also refactored the `I18n` scope of the uploader. Should be all good now, I hope.

tvdeyen commented:
**It now looks like this:**

![upload-overlay](https://cloud.githubusercontent.com/assets/42868/14620404/be0f617c-05bc-11e6-8d0d-09ee870cfcc5.jpg)

![upload-archive](https://cloud.githubusercontent.com/assets/42868/14620405/be33064a-05bc-11e6-8913-ee55b6ff0e0b.jpg)

/c @moritzps @robinboening

mamhoff commented:
🎉
